File: design_oligos.py
#!/usr/bin/env python3
"""
gRNA Cloning Oligo Formatter - Python Version
Formats pre-designed gRNA sequences into cloning-ready oligonucleotides
with appropriate overhangs for BsmBI-based cloning into lentiGuide-puro

NOTE: This tool does NOT design gRNAs. Use gRNA design tools like
CHOPCHOP, Benchling, or CRISPick to design your target sequences first.

Author: Can Ozcan
Date: 2023-07-09
Python Port: 2026-01-28
"""

import pandas as pd
import sys
from pathlib import Path


def reverse_complement(seq):
    """Generate reverse complement of DNA sequence"""
    complement = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}
    return ''.join(complement.get(base, base) for base in reversed(seq))


def design_oligos(input_file='oligo_list.xlsx', output_file='ready_to_order.xlsx'):
    """
    Main function to design CRISPR gRNA oligos
    
    Parameters:
    -----------
    input_file : str
        Path to input Excel file with columns 'name' and 'target_seq'
    output_file : str
        Path to output Excel file
    """
    
    print("=" * 50)
    print("gRNA Cloning Oligo Formatter - Python Version")
    print("=" * 50)
    print()
    
    # Define overhangs
    forw_overhang = "cacc"
    forw_overhang_g = "caccg"
    rev_overhang = "aaac"
    rev_overhang_2 = "c"
    
    # Check if input file exists
    if not Path(input_file).exists():
        print(f"Error: Input file '{input_file}' not found!")
        sys.exit(1)
    
    print(f"Reading input file: {input_file}")
    
    # Read input file
    try:
        df = pd.read_excel(input_file)
    except Exception as e:
        print(f"Error reading file: {e}")
        sys.exit(1)
    
    # Validate columns
    if 'name' not in df.columns or 'target_seq' not in df.columns:
        print("Error: Input file must contain 'name' and 'target_seq' columns")
        sys.exit(1)
    
    print(f"Processing {len(df)} guide sequences...")
    print()
    
    # Process sequences
    results = []
    
    for idx, row in df.iterrows():
        name = row['name']
        target_seq = row['target_seq'].upper()
        
        # Generate reverse complement
        rev_comp = reverse_complement(target_seq)
        
        # Design forward oligo
        if target_seq.startswith('G'):
            forward_oligo = forw_overhang + target_seq
            reverse_oligo = rev_overhang + rev_comp
        else:
            forward_oligo = forw_overhang_g + target_seq
            reverse_oligo = rev_overhang + rev_comp + rev_overhang_2
        
        # Add to results
        results.append({
            'oligo_name': f"{name}-forward_oligo",
            'sequence': forward_oligo
        })
        results.append({
            'oligo_name': f"{name}-reverse_oligo",
            'sequence': reverse_oligo
        })
    
    # Create output dataframe
    output_df = pd.DataFrame(results)
    
    # Clean oligo names (remove _oligo suffix for cleaner names)
    output_df['oligo_name'] = output_df['oligo_name'].str.replace('_oligo', '')
    
    # Write output
    print(f"Writing output to: {output_file}")
    output_df.to_excel(output_file, index=False)
    
    # Print summary
    print()
    print("=" * 50)
    print("Processing complete!")
    print("=" * 50)
    print()
    print("Summary:")
    print(f"- Input sequences: {len(df)}")
    print(f"- Output oligos: {len(output_df)}")
    
    # Count sequences by starting nucleotide
    starts_with_g = df['target_seq'].str.upper().str.startswith('G').sum()
    print(f"- Sequences starting with G: {starts_with_g}")
    print(f"- Sequences NOT starting with G: {len(df) - starts_with_g}")
    
    print()
    print("Output file saved successfully!")
    print(f"File location: {Path(output_file).absolute()}")
    
    # Preview
    print()
    print("First 3 oligo pairs:")
    print(output_df.head(6).to_string(index=False))
    
    return output_df


def main():
    """Command line interface"""
    if len(sys.argv) == 1:
        # Use default filenames
        design_oligos()
    elif len(sys.argv) == 2:
        # Input file specified
        design_oligos(input_file=sys.argv[1])
    elif len(sys.argv) >= 3:
        # Input and output files specified
        design_oligos(input_file=sys.argv[1], output_file=sys.argv[2])
    else:
        print("Usage: python design_oligos.py [input_file] [output_file]")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: design_oligos.R
#!/usr/bin/env Rscript

# gRNA Cloning Oligo Formatter
# Formats pre-designed gRNA sequences into cloning-ready oligonucleotides
# with appropriate overhangs for BsmBI-based cloning into lentiGuide-puro
#
# NOTE: This tool does NOT design gRNAs. Use gRNA design tools like
# CHOPCHOP, Benchling, or CRISPick to design your target sequences first.
#
# Author: Can Ozcan
# Date: 2023-07-09
# Updated: 2026-01-28

# Load required libraries
cat("Loading required packages...\n")
suppressPackageStartupMessages({
  library(readxl)
  library(tidyverse)
  library(openxlsx)
})

# Function to reverse string
strreverse <- function(x) {
  strsplit(x, NULL) %>% 
    lapply(rev) %>% 
    sapply(paste, collapse = "")
}

# Define overhangs for BsmBI-based cloning into lentiGuide-puro
forw_overhang <- "cacc"
forw_overhang_g <- "caccg"
rev_overhang <- "aaac"
rev_overhang_2 <- "c"

# Main function
design_oligos <- function(input_file = "oligo_list.xlsx", 
                          output_file = "ready_to_order.xlsx",
                          sheet_name = "Sayfa1") {
  
  cat("========================================\n")
  cat("gRNA Cloning Oligo Formatter\n")
  cat("========================================\n\n")
  
  # Check if input file exists
  if (!file.exists(input_file)) {
    stop(paste("Error: Input file", input_file, "not found in current directory."))
  }
  
  cat(paste("Reading input file:", input_file, "\n"))
  
  # Read input file
  tryCatch({
    input_data <- read_excel(input_file, sheet = sheet_name)
  }, error = function(e) {
    # If specified sheet doesn't exist, try the first sheet
    cat("Specified sheet not found, trying first sheet...\n")
    input_data <- read_excel(input_file, sheet = 1)
  })
  
  # Validate input columns
  if (!all(c("name", "target_seq") %in% colnames(input_data))) {
    stop("Error: Input file must contain 'name' and 'target_seq' columns")
  }
  
  cat(paste("Processing", nrow(input_data), "guide sequences...\n\n"))
  
  # Generate reverse complement
  input_data <- input_data %>% 
    mutate(reversed = strreverse(target_seq))
  
  input_data$reversed <- chartr("ACGT", "TGCA", input_data$reversed)
  
  # Add appropriate overhangs
  input_data <- input_data %>% 
    mutate(
      forward_oligo = as.character(target_seq),
      forward_oligo = ifelse(
        startsWith(target_seq, "G"),
        paste0(forw_overhang, target_seq),
        paste0(forw_overhang_g, target_seq)
      )
    ) %>% 
    mutate(
      reverse_oligo = as.character(reversed),
      reverse_oligo = ifelse(
        startsWith(target_seq, "G"),
        paste0(rev_overhang, reversed),
        paste0(rev_overhang, reversed, rev_overhang_2)
      )
    )
  
  # Format output
  output_data <- input_data[, c(1, 4, 5)]
  output_data <- output_data %>% 
    pivot_longer(
      !name, 
      names_to = "strand", 
      values_to = "sequence"
    ) %>% 
    mutate(oligo_name = paste0(name, "-", strand))
  
  # Clean up oligo names
  output_data$oligo_name <- gsub("_o.*", "", output_data$oligo_name)
  output_data <- output_data[, c(4, 3)]
  
  # Write output
  cat(paste("Writing output to:", output_file, "\n"))
  write.xlsx(
    output_data, 
    file = output_file, 
    sheetName = "Sheet1", 
    colNames = TRUE, 
    append = FALSE, 
    colWidths = "auto"
  )
  
  cat("\n========================================\n")
  cat("Processing complete!\n")
  cat("========================================\n\n")
  cat("Summary:\n")
  cat(paste("- Input sequences:", nrow(input_data), "\n"))
  cat(paste("- Output oligos:", nrow(output_data), "\n"))
  cat(paste("- Sequences starting with G:", sum(startsWith(input_data$target_seq, "G")), "\n"))
  cat(paste("- Sequences NOT starting with G:", sum(!startsWith(input_data$target_seq, "G")), "\n"))
  cat("\nOutput file saved successfully!\n")
  cat(paste("File location:", file.path(getwd(), output_file), "\n"))
  
  # Preview first few entries
  cat("\nFirst 3 oligo pairs:\n")
  print(head(output_data, 6))
  
  return(invisible(output_data))
}

# Run the main function
if (!interactive()) {
  # Script is being run from command line
  args <- commandArgs(trailingOnly = TRUE)
  
  if (length(args) == 0) {
    # Use default filenames
    design_oligos()
  } else if (length(args) == 1) {
    # Input file specified
    design_oligos(input_file = args[1])
  } else if (length(args) >= 2) {
    # Input and output files specified
    design_oligos(input_file = args[1], output_file = args[2])
  }
} else {
  cat("Script loaded. Run design_oligos() to process your data.\n")
  cat("Usage: design_oligos(input_file = 'oligo_list.xlsx', output_file = 'ready_to_order.xlsx')\n")
}
